src/content/lib/ua_overrider.js

```diff
@@ -0,0 +1,83 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const {Cc, Ci} = require("chrome");
+
+const ObserverService = Cc["@mozilla.org/observer-service;1"].getService(Ci.nsIObserverService);
+const DefaultUA = Cc["@mozilla.org/network/protocol;1?name=http"].getService(Ci.nsIHttpProtocolHandler).userAgent;
+const eTLDService = Cc["@mozilla.org/network/effective-tld-service;1"].getService(Ci.nsIEffectiveTLDService);
+
+const NS_HTTP_ON_USERAGENT_REQUEST_TOPIC = "http-on-useragent-request";
+
+class UAOverrider {
+  constructor(overrides) {
+    this._overrides = {};
+    this._overrideForURICache = new Map();
+
+    this.initOverrides(overrides);
+  }
+
+  initOverrides(overrides) {
+    for (let override of overrides) {
+      if (!this._overrides[override.baseDomain]) {
+        this._overrides[override.baseDomain] = [];
+      }
+
+      if (!override.uriMatcher) {
+        override.uriMatcher = () => true;
+      }
+
+      this._overrides[override.baseDomain].push(override);
+    }
+  }
+
+  init() {
+    ObserverService.addObserver(this, NS_HTTP_ON_USERAGENT_REQUEST_TOPIC, false);
+  }
+
+  uninit() {
+    ObserverService.removeObserver(this, NS_HTTP_ON_USERAGENT_REQUEST_TOPIC);
+  }
+
+  observe(subject, topic) {
+    if (topic !== NS_HTTP_ON_USERAGENT_REQUEST_TOPIC) {
+      return;
+    }
+
+    let channel = subject.QueryInterface(Ci.nsIHttpChannel);
+    let uaOverride = this.getUAForURI(channel.URI);
+
+    if (uaOverride) {
+      channel.setRequestHeader("User-Agent", uaOverride, false);
+    }
+  }
+
+  getUAForURI(uri) {
+    let bareUri = uri.specIgnoringRef;
+    if (this._overrideForURICache.has(bareUri)) {
+      let cachedUA = this._overrideForURICache.get(bareUri);
+      return (cachedUA ? cachedUA : DefaultUA);
+    }
+
+    let finalUA = this.lookupUAOverride(uri);
+    this._overrideForURICache.set(bareUri, finalUA);
+
+    return finalUA;
+  }
+
+  lookupUAOverride(uri) {
+    let baseDomain = eTLDService.getBaseDomain(uri);
+    if (this._overrides[baseDomain]) {
+      for (let uaOverride of this._overrides[baseDomain]) {
+        if (uaOverride.uriMatcher(uri.specIgnoringRef)) {
+          return uaOverride.uaTransformer(DefaultUA);
+        }
+      }
+    }
```

Comment:
For my test case (baseDomain: 163.com, test URI: http://sports.163.com), I can't intercept all network connects, like http://goal.sports.163.com/teamlogo/7495.png .